Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+        switch (address.address_type()) {
+        case AddressType::Line:
+            builder.appendff("{}", address.line_number());
+            break;
+        case AddressType::LastLine:
+            builder.append('$');
+            break;
+        case AddressType::ContextAddress:
+            VERIFY_NOT_REACHED();
+        case AddressType::Unset:
+            break;
+        }
+    }
+};
+}
+
+struct Command {
+    Address addr1;
+    Address addr2;
+    char function = '\0';
+    StringView arguments;
+
+    void enable_for(StringView pattern_space, size_t line_number, bool is_last_line)
+    {
+        m_is_enabled = selects(pattern_space, line_number, is_last_line);
+    }
+
+    bool is_enabled() const { return m_is_enabled; }
+
+private:
+    bool m_is_enabled { false };
+    bool m_is_selection_active { false };
+
+    bool selects(StringView pattern_space, size_t line_number, bool is_last_line)
+    {
+        // no address set, all patterns match
+        if (addr1.address_type() == AddressType::Unset) {
+            VERIFY(addr2.address_type() == AddressType::Unset);
+            return true;
+        }
+
+        // single address set
+        if (addr2.address_type() == AddressType::Unset)
+            return addr1.matches(pattern_space, line_number, is_last_line);
+
+        // two addresses
+        if (!m_is_selection_active && addr1.matches(pattern_space, line_number, is_last_line)) {
+            m_is_selection_active = true;
+            return true;
+        } else if (m_is_selection_active) {
+            if (addr2.matches(pattern_space, line_number, is_last_line))
+                m_is_selection_active = false;
+            return true;
+        }
+        return false;
+    }
+};
+
+namespace AK {
+template<>
+class Formatter<Command> : public StandardFormatter {
+public:
+    void format(FormatBuilder& format_builder, Command command)
+    {
+        auto& builder = format_builder.builder();
+        builder.appendff("{}", command.addr1);
+        if (command.addr2.address_type() != AddressType::Unset) {
+            builder.appendff(",{}", command.addr2);
+        }
+        builder.append(command.function);
+        builder.append(command.arguments);
+    }
+};
+}
+
+struct AddressParsingResult {
+    Optional<Address> address;
+    size_t end_of_address;
+};
+
+static AddressParsingResult parse_address(StringView sv)
+{
+    if (sv.is_empty())
+        return { {}, 0 };
+
+    if (sv[0] == '$') {
+        return { Address { AddressType::LastLine }, 1 };
+    }
+
+    size_t index = 0;
+    for (; index < sv.length() && '0' <= sv[index] && sv[index] <= '9'; index++)
+        ;
+    if (index == 0)
+        return { {}, 0 };
+    return {
+        Address { AK::StringUtils::convert_to_uint<size_t>(sv.substring_view(0, index)).release_value() },
+        index
+    };
+}
+
+template<int max_addresses>
+static bool verify_number_of_addresses(const Command& command)
+{
+    if constexpr (max_addresses == 2)
```

Comment:
As the `else` block has braces, please put braces around the then block too.

Reply:
Done